datafusion/expr/src/logical_plan/tree_node.rs

```diff
@@ -32,23 +39,368 @@ impl TreeNode for LogicalPlan {
         self.inputs().into_iter().apply_until_stop(f)
     }
 
-    fn map_children<F: FnMut(Self) -> Result<Transformed<Self>>>(
-        self,
-        f: F,
-    ) -> Result<Transformed<Self>> {
-        let new_children = self
-            .inputs()
-            .into_iter()
-            .cloned()
-            .map_until_stop_and_collect(f)?;
-        // Propagate up `new_children.transformed` and `new_children.tnr`
-        // along with the node containing transformed children.
-        if new_children.transformed {
-            new_children.map_data(|new_children| {
-                self.with_new_exprs(self.expressions(), new_children)
-            })
-        } else {
-            Ok(new_children.update_data(|_| self))
-        }
+    /// Applies `f` to each child (input) of this plan node, rewriting them *in place.*
+    ///
+    /// # Notes
+    ///
+    /// Inputs include ONLY direct children, not embedded `LogicalPlan`s for
+    /// subqueries, for example such as are in [`Expr::Exists`].
+    fn map_children<F>(self, mut f: F) -> Result<Transformed<Self>>
+    where
+        F: FnMut(Self) -> Result<Transformed<Self>>,
+    {
+        Ok(match self {
```

Comment:
This follows the (really very cool) pattern @peter-toth came up with in https://github.com/apache/arrow-datafusion/pull/9913